src/Analyzers/Core/Analyzers/MakeFieldReadonly/AbstractMakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -234,11 +234,26 @@ static bool IsDataContractSerializable(IFieldSymbol symbol, INamedTypeSymbol? da
     private static Location GetDiagnosticLocation(IFieldSymbol field)
         => field.Locations[0];
 
+    private static bool IsWrittenTo(IOperation? operation, ISymbol owningSymbol)
+    {
+        if (operation is null)
+            return false;
+
+        var result = operation.GetValueUsageInfo(owningSymbol);
+        if (result.IsWrittenTo())
+            return true;
+
+        // Accessing an indexer/property off of a value type will read/write the value type depending on how the
+        // indexer/property itself is used.
+        return operation is { Type.IsValueType: true, Parent: IPropertyReferenceOperation }
+            ? IsWrittenTo(operation.Parent, owningSymbol)
+            : false;
```

Comment:
I guess an extra check here that the property reference is not readonly may make sense.

```
public struct S
{    
    public string P1 { get; set; }

    public readonly string P2 { get => ""; set { } }
}

public class C
{
    private readonly S _s;
    
    
    public void M()
    {
        _s.P1 = ""; // not valid as _s is read-only and this is considered a write
        _s.P2 = ""; // perfectly valid as the property is readonly.
    }
}
```

Reply:
@tannergooding This comment is on the "make field readonly" analyzer, unrelated to the "unnecessary assignment".

Currently there is a small false negative where we won't suggest making a field readonly if a readonly property setter is accessed. Based on internal discussion with Cyrus, it's not an important scenario and we are okay with the false negative.

Reply:
> and we are okay with the false negative.

Yup. False negatives are usually totally ok.

True negatives/positives are what we desire all the time. 

False positives are what we try hard to avoid. But *occasionally* will accept if the overall value to the user is high, and the chance of issue is low. For example offering to change `x.Prop != null ? x.Prop : y.Prop` into `x.Prop ?? y.Prop`. This is technically a false positive as the semantics do change here. But the overall value in detecting this, and the *extremely low* rate of this changing semantics in a negative way for a user outweighs the feature to avoid the false positive.